lib/dom_utils.coffee

```diff
@@ -131,21 +130,69 @@ DomUtils =
 
     null
 
-  #
-  # Bounds the rect by the current viewport dimensions. If the rect is offscreen or has a height or width < 3
-  # then null is returned instead of a rect.
-  #
-  cropRectToVisible: (rect) ->
-    boundedRect = Rect.create(
-      Math.max(rect.left, 0)
-      Math.max(rect.top, 0)
-      rect.right
-      rect.bottom
-    )
-    if boundedRect.top >= window.innerHeight - 4 or boundedRect.left >= window.innerWidth - 4
-      null
-    else
-      boundedRect
+  # Bounds the rect by the current viewport dimensions. If the rect is offscreen or has a visible height or
+  # width <= 3 then null is returned instead of a rect.
+  Viewport: do ->
+    getMinX = -> 0
+    getMinY = -> 0
+    getMaxX = -> window.innerWidth
+    getMaxY = -> window.innerHeight
+
+    # Inform frames and iframes of the visible portion of their window; for example, an iframe may be
+    # partially or wholly outside of the viewport.
+    postTimer = null
+    onScroll = (event) ->
+      clearTimeout postTimer if postTimer
+      postTimer = Utils.setTimeout 50, ->
+        postTimer = null
+        chrome.storage.local.get "vimiumSecret", ({vimiumSecret: secret}) ->
+          elements = []
+          elements.push element for element in document.documentElement.getElementsByTagName "iframe"
+          elements.push element for element in document.documentElement.getElementsByTagName "frame"
+          for element in elements
+            rect = do ->
+              rect = element.getBoundingClientRect()
+              left = Math.max 0, -rect.left
+              top = Math.max 0, -rect.top
+              right = Math.max 0, Math.min rect.width, window.innerWidth - rect.left
+              bottom = Math.max 0, Math.min rect.height, window.innerHeight - rect.top
+              right = left if right < left
+              bottom = top if bottom < top
+              Rect.create left, top, right, bottom
+            element.contentWindow.postMessage {name: "vimiumRegisterFrameViewport", rect, secret}, "*"
```

Comment:
* Don't send `vimiumSecret` to a frame that has non-Vimium code running in it. Otherwise, a malicious sub-frame can open and control any of the `UIComponent` pages.
* Don't use `window.postMessage`:
  - pages with `message` listeners may break when we send messages
  - pages who want to mess with Vimium can send themselves/child frames a message misreporting their size

As an alternative, you could try
```coffee
event = new CustomEvent "vimiumRegisterFrameViewport-#{secret}", { detail: { rect } }
element.contentWindow.dispatchEvent event
```
and attach the corresponding listener on each page.

Reply:
> Don't send vimiumSecret to a frame that has non-Vimium code running in it. 

Doh.  Yes.  Thanks.

> As an alternative, you could try .... (a custom event)

I had tried that and just double checked.  Unfortunately it triggers a cross-origin frame security error.

![image](https://user-images.githubusercontent.com/2641335/34917241-9b80bd44-f93b-11e7-907d-12480734a000.png)

> pages who want to mess with Vimium can send themselves/child frames a message misreporting their size

I'd be less worried about that (there's nothing for such a site to gain).  For example, a site can already mess with Vimium by removing the Vomnibar frame.

But...

> pages with message listeners may break when we send messages

How much of a concern is that really?  A page is pretty broken if sending it an arbitrary message breaks it.  I'd put this in the *let's try it out and see* class of issues.